cmake/Version.cmake

```
@@ -14,6 +16,81 @@ ly_file_read("${LY_ROOT_FOLDER}/engine.json" tmp_json_data)
 set_property(GLOBAL PROPERTY O3DE_ENGINE_JSON_DATA ${tmp_json_data})
 unset(tmp_json_data)
 
+#! o3de_get_name_and_version_specifier: Parse the dependency name, and optional version 
+# operator and version from the supplied input string. 
+# \arg:input - the input string e.g. o3de==1.0.0
+# \arg:dependency_name - the name to the left of any optional version specifier
+# \arg:operator - the version specifier operator e.g. == 
+# \arg:version - the version specifier version e.g. 1.0.0
+function(o3de_get_name_and_version_specifier input dependency_name operator version)
+    if("${input}" MATCHES "^(.*)(~=|==|!=|<=|>=|<|>|===)(.*)$")
+        if(${CMAKE_MATCH_COUNT} GREATER_EQUAL 1)
+            set(${dependency_name} ${CMAKE_MATCH_1} PARENT_SCOPE)
+        endif()
+        if(${CMAKE_MATCH_COUNT} GREATER_EQUAL 2)
+            set(${operator} ${CMAKE_MATCH_2} PARENT_SCOPE)
+        endif()
+        if(${CMAKE_MATCH_COUNT} GREATER_EQUAL 3)
+            set(${version} ${CMAKE_MATCH_3} PARENT_SCOPE)
+        endif()
```

The match count here can never be less than 4.
The regular expression cannot succeed in any other way.
A capture 0 is implicit and capture 1-3 aren't optional, that must be their.
They capture group may be an empty string but they must be there.
Also there is a flaw with the regular expression
It can no characters on either size of the relational operator
i.e "o3de==" will match, "==1.0.0" will match as well.
Second problem if for some reason the user specifies the version specifier as `o3de == 1.0.0` that will also fail because the `dependency_name` and `version` values don't strip surrounding whitespace.
```suggestion
    if("${input}" MATCHES "^(.+)(~=|==|!=|<=|>=|<|>|===)(.+)$")
        string(STRIP ${CMAKE_MATCH_1} _dependency_name)
        set(${dependency_name} ${_dependency_name} PARENT_SCOPE)
        set(${operator} ${CMAKE_MATCH_2} PARENT_SCOPE)
        string(STRIP ${CMAKE_MATCH_3} _version)
        set(${version} ${_version} PARENT_SCOPE)
```

---

In my tests the old regex looks like it works OK and maybe handles one case better (where version is missing) - except for the space stripping, thoughts? https://godbolt.org/z/3dnqqod9a

```
OLD 'o3de == 1.0.0' object_name:'o3de' op:'==' ver: '1.0.0'
NEW 'o3de == 1.0.0' object_name:'o3de' op:'==' ver: '1.0.0'

OLD 'o3de>=1.2.3' object_name:'o3de' op:'>=' ver: '1.2.3'
NEW 'o3de>=1.2.3' object_name:'o3de' op:'>=' ver: '1.2.3'

OLD 'o3de' object_name:'o3de' op:'' ver: ''
NEW 'o3de' object_name:'o3de' op:'' ver: ''

OLD 'o3de>=' object_name:'o3de' op:'>=' ver: ''
NEW 'o3de>=' object_name:'o3de' op:'>' ver: '='
```